keps/sig-scheduling/5598-scheduling-cache/README.md

```diff
@@ -0,0 +1,955 @@
+
+## Motivation
+
+Today our scheduling algorithm is O(num pods x num nodes) where num pods is the number of pods to schedule
+and num nodes the number of nodes in the cluster. As the size of clusters and jobs continue
+to increase, this leads to low performance when scheduling or rescheduling 
+large jobs. This increases customer cost and slows down customer jobs, both
+unpleasant impacts. Optimizations like this one have the potential to dramaticly
+reduce the cost of scheduling in these scenarios.
+
+We are also working on multi-pod scheduling, which will give us a way to consider multiple 
+pods at the same time. While full multi-pod scheduling will likely deprecate the cache, 
+the signatures work and structuring of the cache dependencies should be portable to the new cycles
+that consider multiple pods.
+
+Another change is the shift towards 1-pod-per-node in batch and ML environments. Many
+of these environments (among others) only attempt to run a single customer pod on each node, along
+with a complement of daemon set pods. This simplifies our scheduling needs significantly.
+
+### Goals
+
+ * Improve the performance of scheduling large jobs on large clusters where the constraints are simple.
+ * Begin building infrastructure to support gang scheduling and other "multi-pod" scheduling requests.
+ * Ensure that the infrastructure we build is maintainable as we update, add and remove plugins.
+ * Never impact feasibility.
+ * Provide identical results to our current scheduler for 1-pod-per-node environments.
+ * Provide an experience with limited changes in multi-pod-per-node environments.
+
+### Non-Goals
+
+ * We are not attempting to apply this optimization to all pods. Some constraints
+do not lend themselves to caching. Instead, we focus on customers who have
+large jobs, who need high performance, and do not generally use these
+more complex constraints. While we expect the cache to always be on, many pods will not be 
+able to take advantage of it.
+ * We are not adding gang scheduling of any kind in this KEP. This is purely a performance
+improvement, although we hope the work on this KEP will help us with gang scheduling as we build it.
+
+## Proposal
+
+See https://github.com/bwsalmon/kubernetes/pull/1 for a draft version of the code. Feedback is extremely welcome.
+
+We introduce two things:
+ * The pod cache.
+ * A pod scheduling signature
+
+### Pod cache
+
+The end result of the scheduler cycles is a list of node names that will fit
+the current pod, ordered by their score from most to least attractive. This result
+is what we cache; the current pod will use the first node on the list, and we cache the remaining
+results, indexed by the pod's "scheduling signature" (which we will describe later).
+
+When a pod with the same signature comes later, we find the entry in our cache and pull the first
+node off the list. We then go down the nominated node path. Just as we would with a pod
```

Comment:
So, there's multiple lists, one per one signature. When a scheduling happens and node#1 is selected, are we going to iterate all lists and remove node#1 from all lists? (assuming node#1 is no longer available for pods because of 1-pod-per-node assumption)